Ember upgrade v3.20 to v3.28 with lay-out improvements

benjay10 opened the pull request:
This PR is based around an overdue Ember upgrade from version 3.20 to version 3.28 LTS. Most other external packages have also been upgraded at once.

During this process some legacy components stopped working, so most of the layout of this app has been revised with new ember-appuniversum components. This PR features a new ember-data-table, removed old styling and replaced by utility classes, loading screens, error screen, and entirely redesinged edit page with a sidebar.

Some work has also been done to fix some (spelling) errors in model properties and some minor updates to use the new cogs:Job model.

benjay10 committed:
Bumped external package versions

These package bumps include only packages not related to Ember as those
packages will be managed by the ember-cli-update later.

benjay10 committed:
Enabled the package-lock.json file

This is usefull and should be enabled. Escpecially for debugging package
problems and to deploy a stable build.

benjay10 committed:
Cleaned up styling on mock-login

The mock-login was not displaying properly anymore. These changes
include a copy from another frontend where the mock-login was also used.

benjay10 committed:
Ember upgrade v3.20.2...v3.28.5

This is the result of using ember-cli-update.

benjay10 committed:
Move to npm 8 and resolve peerdependency issue

Package @lblod/ember-submission-form-fields required an older version of
ember-power-select while a newer version was installed. This came to
light by using npm 8 with a more thourough peerdependency check
mechanism.

benjay10 committed:
Resolved deprecation warnings

To resolve some of these warnings, a class had to be rewritten using the
modern Octane style of components in Ember. Also some config on
ember-simple-auth had to be revised.

benjay10 committed:
Repaired constand logging out due to error

Due to an error in the CurrentSessionService, the application kept
logging out.

benjay10 committed:
Fixed deprecations warning about transitionTo

benjay10 committed:
Fixed deprecation about AuToolBarGroup and LinkRoute

benjay10 committed:
Automatic linting fixes and removal of tests

All the test were automatically generated and can be removed as they
don't really help at all. They can easily be generated again if needed.

benjay10 committed:
Updated code header and logout design

benjay10 committed:
Modal from task to job, reviewed forms table

A submission does not longer refer to an automatic submission task, but
to a job. This is now reflected in the models. Also rewrote the data
table to list the forms to make use of the new Appuniversum components.

benjay10 committed:
Replace moment by date-fns

Moment is being phased out in favour of date-fns, which is also used in
ember-appuniversum. Made a helper function to format dates and used that
throughout.

benjay10 committed:
Redesign of the form edit page

This redesign uses more vertical space on wider screens. It should also
scale to smaller screens while still being functional.

benjay10 committed:
Revamped the edit page

It now comes with information on the left, and two tabs to separate the
heavy tables from the from editing. This is done using visibility
classes instead of removing/adding component bacause of issues with the
form component.

benjay10 committed:
Refreshed loading pages, added global loading page

benjay10 committed:
Simplified all the alert

There were just too many alerts, this has now been simplified using
toasters, which survive route transitions.

benjay10 committed:
Removed package ember-vo-mu-file-upload

This package was not used, but probably remained bacause of a old
peerdependency problem.

benjay10 committed:
Add error page with error message and code

This is also an implicit route that only needs a template page, similar to
the loading substate.

benjay10 committed:
Removed style that is now part of AU or redundant

Most of these deletes are because that style is already present in
AppUniversum, or because these style rules are no longer needed since
they are featured in a more recent version of AU.

benjay10 committed:
Reviewed model classes and synced with resource

There were some spelling mistakes in the models, also some things have
changed, such as the submission to job relationship.

benjay10 committed:
Loading screen only on mock-login loading

The index page should have its own loading mechanism, a spinner in the
datatable. The edit and new pages also have their own loading page.

benjay10 committed:
Correct includes for less requests

With the necessary includes, the model request for the ember-data-table
is larger but faster. The data also renders at once instead of in little
chunks.

benjay10 committed:
Fixed styling after removed rule on type text

benjay10 committed:
Clearer link to open submission on main page

benjay10 committed:
Added translations

Installed ember-intl and wrote translations for the whole app.
During my testing, the app got stuck on English. I hope this is not a
permanent thing as I could not find a place where the locale was set to
anything fixed.

benjay10 committed:
Correctly detect users locale

Turns out you have to manually set the locale for the application. This
is now done by passing the language set in the browser, and defaulting
to nl-BE.

benjay10 committed:
Lightly updated the README with some more specifics

Windvis commented on app/components/triples-table.hbs:
```diff
@@ -1,20 +1,20 @@
 {{#if @triples}}
-  <table class="triple">
-    <thead>
+  <AuTable @size="default">
```
nitpick, size can be dropped
```suggestion
  <AuTable>
```

cecemel commented:
@benjay10 I tested, seems nice!

Windvis commented on app/helpers/date-format.js:
```diff
@@ -0,0 +1,10 @@
+import { helper } from '@ember/component/helper';
+import { format } from 'date-fns';
```
I'm not sure, but I think this will bundle the whole `date-fns` instead of only the format function. Last time I checked this required a nested path to properly treeshake, but that might have changed.

benjay10 replied:
I don't know what you mean. Can we bundle only a function instead of the whole package? Do you mean something similar as the `lodash.somefunction` packages? I don't know if date-fns supports that.

I tried something like `import { format } from 'date-fns/format';` but that doesn't work.

Windvis replied:
Yea, Webpack can treeshake unused functions, but it can't for some reason if you import from the top level package.

I think it's a default export so like this: `import format from 'date-fns/format'`.

Windvis replied:
You can check quickly by doing a production build and comparing the size changes. Maybe the problem has been fixed and the subpath import isn't needed anymore.

Windvis replied:
For reference, this is the underlying issue that prevents it from working without the subpath: https://github.com/ef4/ember-auto-import/issues/121

So until that is fixed, or we switch to Embroider, the subpath is needed to reduce the bundle size.

benjay10 replied:
The package size difference between the two build (one with subpath and one without) is 75.632 bytes. So I will use the subpath method.

Windvis commented on app/routes/application.js:
```diff
@@ -1,28 +1,28 @@
 
   sessionInvalidated() {
-    this.transitionTo('mock-login');
-  },
+    this.router.transitionTo('login');
+  }
```
This is no longer used in the new setup since the ApplicationRouteMixin is gone. The session service should be extended and it has a hook that can be used.

benjay10 replied:
I'll just remove this function, invalidation of the session is done correctly without this. The login route redirects to the 'mock-login' route.

Windvis commented on app/routes/index.js:
```diff
@@ -1,15 +1,31 @@
+import { inject as service } from '@ember/service';
+
+export default class IndexRoute extends Route.extend(DataTableRouteMixin) {
+  @service session;
```
I think the store service also needs to be injected since the mixin uses it internally.

benjay10 replied:
Aah finally! I could not find where these deprecation warnings came from.

Windvis commented on app/routes/mock-login.js:
```diff
@@ -1,27 +1,27 @@
-    if (this.session.isAuthenticated)
-      this.transitionTo('index');
-  },
+    if (this.session.isAuthenticated) this.transitionTo('index');
```
Nitpick: This can be replaced with [`session.prohibitAuthentication`](http://ember-simple-auth.com/api/classes/SessionService.html#method_prohibitAuthentication)

Windvis commented on app/services/current-session.js:
```diff
@@ -1,60 +1,43 @@
+        'account',
+        this.session.data.authenticated.relationships.account.data.id
+      );
+      this.user = await this.account.get('gebruiker');
```
`get` shouldn't be needed if you await it.
```suggestion
      this.user = await this.account.gebruiker;
```

Windvis commented on app/services/current-session.js:
```diff
@@ -1,60 +1,43 @@
-      });
+  async load() {
+    if (this.session.isAuthenticated) {
+      this.account = await this.store.find(
```
store.find is a private API so I would replace it with the `findRecord` equivalent. It works exactly the same but `.find` is only there to be compatible with some Ember magic behavior.
```suggestion
      this.account = await this.store.findRecord(
```

Windvis commented on app/services/current-session.js:
```diff
@@ -1,60 +1,43 @@
+        'bestuurseenheid',
+        this.session.data.authenticated.relationships.group.data.id
+      );
+      this.roles = await this.session.data.authenticated.data.attributes.roles;
```
`await` isn't needed here I think since it's just an object property lookup.

Windvis commented on app/styles/_shame.scss:
```diff
@@ -131,3 +63,47 @@ $au-non-editable-background    : $au-gray-100 !default;
+  Pagination on the mock-login (possibly among others)
+******************************************************************************/
+
+.data-table-pagination {
```
The Appuniversum equivalent can be used instead.

https://github.com/lblod/frontend-loket/blob/29d5e09748320dcdc75f0971c7d5c8078a8e7e6b/app/templates/mock-login.hbs#L39-L45

benjay10 replied:
I saw that, but it doesn't provide the same functionality. I would miss the select for the pages. When logging in for the "Gemeente Mechelen" I know I have to go to page 3 in one action. This can't be done with the new component.

I will change it to the component you suggest, because that is even less styling and it is a supported component by ember-appuniversum.

Windvis replied:
Yea agreed, but since it's only for the mock login I don't think it matters that much. I'm not sure why the page selector isn't present in the Appuniversum equivalent.

Windvis commented on app/templates/index.hbs:
```diff
@@ -1,36 +1,69 @@
-          @size={{this.size}}
-          as |t|
-  >
+{{#let (t "main.title")
+       (t "main.create-submission")
+       (t "submission.unit")
+       (t "submission.type")
+       (t "submission.date")
+       (t "submission.creator")
+       (t "submission.modifier")
+       (t "submission.sent")
+       (t "submission.status")
+       (t "submission.action")
+       (t "submission.created-automatic")
+       (t "submission.details")
```
Nitpick, but I find this strange to see and it looks like a pain to maintain since it's order dependent. Only `createdAutomatic` seems used twice, but even then I would just write it twice instead of forcing an extra level of indentation 😄.

benjay10 replied:
I really despise the way this looks. The let is needed because for some reason the DataTable and those translation don't want to mix.

Indentation? Every item on a new line makes it more readable, no? Also much easier to edit for me using block editing.

Windvis replied:
The reason it doesn't work in the current code is because the block param is also called `t` so it shadows the helper. If that is renamed to something like `table` it will work again.

With indentation I meant that the `let` content is indented an extra level.

benjay10 replied:
Ah another one of those "how did I not see that before" things. I'll clean this mess up.

And the indentation is because block content is always indented for me. If blocks, unless blocks, HTML block content, ... so also the let block content.

Windvis commented on .npmrc:
```diff
@@ -1 +1 @@
-package-lock=false
\ No newline at end of file
+package-lock=true
```
I would just remove the file since this is the default.

Windvis replied:
Also, this seems to be a v2 lock file so I think the Dockerfile and .drone.yml need to be updated to a version that ships with npm 8. And npm CI can  be used to speed up the install (which would also require copying the lock file into the build container).

benjay10 replied:
File removed, need to look into the drone config still.

Windvis reviewed (commented):
This was a _hard_ review 😄. Please split it up in multiple PRs in the future 🙏.

Thanks for the maintenance work!

benjay10 committed:
Update app/components/triples-table.hbs

Removed default value.

Co-authored-by: Sam Van Campenhout <sam.van.campenhout@gmail.com>

benjay10 committed:
Removed unused function about session invalidation

benjay10 committed:
Resolve deprecation warning on store injection

benjay10 committed:
Use nicer session.prohibitAuthentication

benjay10 committed:
Update app/services/current-session.js

Using `.get()` is not needed here.

Co-authored-by: Sam Van Campenhout <sam.van.campenhout@gmail.com>

benjay10 committed:
Update app/services/current-session.js

Use findRecord instead of find. This should be the newer api.

Co-authored-by: Sam Van Campenhout <sam.van.campenhout@gmail.com>

benjay10 committed:
Removed await in currentSession

This one wasn't needed indeed.

benjay10 committed:
Removed .npmrc as it only contained a default value

benjay10 committed:
Changed NumberPagination to AuNumberPagination

...and removed unused styling as a consequence.

benjay10 committed:
Merge branch 'ember-upgrade-v3.20-v3.28' of github.com:lblod/frontend-meldingsplichtige-api into ember-upgrade-v3.20-v3.28

benjay10 commented:
All 7 commits above are thanks to the review of @Windvis! Thanks for that!

benjay10 committed:
Fixed deprecation warning on missing router service

benjay10 committed:
Making sure the user and groups are loaded on create

Apparently, without this change, after a fresh login, you could not
create a manual submission. This was due to an `undefined` user and
group in the request to mu-cl-resource. It would create resources with
missing data that then causes bigger issues later on while editing the
submission.

benjay10 commented:
@Windvis Can you check out [this commit (the one before this comment)](https://github.com/lblod/frontend-meldingsplichtige-api/pull/12/commits/a5d3886d433ebf073e9416670a6320cf9c7a846a)? Why is this necessary?

Felix and I have both noticed that creating a manual submission fails after a fresh login. The request sent to mu-cl-resources does not contain the creator and bestuurseenheid as it should. Logging the request objects reveals the user and group to be `undefined`. When performing a call to `load()` on the current-session service, everything works as normal every time.

Windvis commented:
> @Windvis Can you check out [this commit (the one before this comment)](https://github.com/lblod/frontend-meldingsplichtige-api/pull/12/commits/a5d3886d433ebf073e9416670a6320cf9c7a846a)? Why is this necessary?

> Felix and I have both noticed that creating a manual submission fails after a fresh login. The request sent to mu-cl-resources does not contain the creator and bestuurseenheid as it should. Logging the request objects reveals the user and group to be undefined. When performing a call to load() on the current-session service, everything works as normal every time.

Oh, right. It's related to the new ember-simple-auth setup. At the moment the currenSession is only loaded when the application boots (since the application route calls it). Clicking a user in the mock-login route doesn't force a refresh, so the load method isn't being called anywhere.

In the previous implementation this was handled by the application route as well. The `ApplicationRouteMixin` called the `sessionAuthenticated` method on the route if the session was authenticated, but that no longer happens. The equivalent in the new setup is to create a new SessionService which extends the one from the addon and implement the correct hook.

The loket implementation for reference: https://github.com/lblod/frontend-loket/blob/bf88ecee49e0bcf6f3b0ca4739427f50b819bf33/app/services/session.js#L9-L12

cecemel commented on app/routes/forms/new.js:
```diff
@@ -3,22 +3,31 @@ import { inject as service } from '@ember/service';
   }
 
   async model() {
+    await this.currentSession.ensureLoaded();
```
@benjay10 @Windvis I am not sure this is needed. We don't do this in loket. I would expect it to be fixed on level of the application route

Windvis replied:
Yea, once the session service hook is implemented this won't be needed anymore.

cecemel replied:
ok! sorry  missed part of the review
@benjay10 this is also the reason why such flaky behavior when creating new submission. The bestuurseenheid is sometimes empty, hence everything else crashes in enrich submissions

benjay10 replied:
@cecemel Yes that is exactly what I wrote about in the commit message and the above comment.

benjay10 replied:
@Windvis Thanks, this works!

benjay10 committed:
Remove `let` for translations

Removing the `let` that was only used for translations cleans up the
code. It was there in the first place because I didn't see the name
clash on `t`.

benjay10 committed:
Use subpath import for tree shaking

This allows for smaller package build sizes.

benjay10 committed:
Implemented sublcass of session service

This handles authentication better and loads the user information at the
correct time.

benjay10 committed:
Updated drone config for new version of npm

Also used `npm ci` now that the package-lock.json file has been enabled.
Included ember-cli-dependency-lint and a script for it.

Windvis commented on package.json:
```diff
@@ -14,77 +14,86 @@
+    "ember-cli-app-version": "^5.0.0",
+    "ember-cli-babel": "^7.26.10",
     "ember-cli-dependency-checker": "^3.2.0",
+    "ember-cli-dependency-lint": "^2.0.1",
```
I would also add the config to not generate any tests since we already run the cli command.

https://github.com/salsify/ember-cli-dependency-lint#lint-tests

benjay10 replied:
I don't know what this means. I don't see how this is generating tests.

Windvis replied:
At the moment ember-cli-dependency-lint generates a test for each addon in the app. Since we already run the cli equivalent with the `lint:dependencies` script we actually run the dependency lint step twice (as can be seen in the CI logs: https://rpio-drone.redpencil.io/lblod/frontend-meldingsplichtige-api/29/1/6).

The addon provides an option to disable that which is documented in the link of my previous comment 😄. So just adding that config should be all that's needed.

Windvis commented on app/services/session.js:
```diff
@@ -0,0 +1,11 @@
+import { inject as service } from '@ember/service';
+import SessionService from 'ember-simple-auth/services/session';
+
+export default class LoketSessionService extends SessionService {
```
Copy pasted? 😄
```suggestion
import BaseSessionService from 'ember-simple-auth/services/session';

export default class SessionService extends BaseSessionService {
```

benjay10 replied:
Yes of course I copy-pasted. :roll_eyes:

Windvis commented:
~~The [Dockerfile](https://github.com/lblod/frontend-meldingsplichtige-api/blob/1b2fc703ff9e040083e2287670b81c8859d01463/Dockerfile#L1-L7) still needs to be updated I think?~~

benjay10 committed:
Update app/services/session.js

Co-authored-by: Sam Van Campenhout <sam.van.campenhout@gmail.com>

benjay10 committed:
Updated Dockerfile with new ember, proxy and npm ci

benjay10 committed:
Merge branch 'ember-upgrade-v3.20-v3.28' of github.com:lblod/frontend-meldingsplichtige-api into ember-upgrade-v3.20-v3.28

benjay10 committed:
Disable dependency tests on addons

benjay10 committed:
Also copy package-lock.json file for Docker build

This is in the Drone config
